uktrade/directory-sso
# Feature greatuk 1270 add get request caching

This PR is to implement GET request caching in directory-sso and cache invalidation into POST, and PUT requests.

Testing:

1) Open a new browser tab, visit http://greatcms.trade.great:8020/signup/ and create a new user. This will make all the necessary calls to directory-sso i.e. create a new User and UserProfile object.

2) Create a superuser in your directory-sso instance.

3) Open an incognito browser tab and visit http://sso.trade.great:8003/api/v1/user/cache/
You should have acces to django-debug-toolbar and see in the cache tab that the GET request has missed the cache:
<img width="1133" alt="Screenshot 2024-10-29 at 16 21 05" src="https://github.com/user-attachments/assets/b6cb0dd0-9206-41ba-a9da-288392ea2691">

This GET request has now cached the data for 30 minutes. If you refresh your browser and check again, you will see that the request hits the cache.

<img width="1129" alt="Screenshot 2024-10-29 at 16 21 22" src="https://github.com/user-attachments/assets/c9ec783a-4928-4660-a5da-366a2591b045">

4) Re open your browser tab where you created a new user. Visit http://greatcms.trade.great:8020/profile/personal-profile/ and add/update some details. This will make a PUT request to directory-sso and clear the cache.

Your directory-sso terminal will look like the following image. You will see that we have 2 logs.
The first says "A models.Model.save() has been invoked - attempting cache.clear() call.". The second says "cache.clear() call complete.".
<img width="866" alt="Screenshot 2024-10-30 at 09 45 00" src="https://github.com/user-attachments/assets/7c6f8eac-24df-4318-ae12-00593c9b3046">


5) re open your incognito browser and refresh the screen. You will see that the cache has now been cleared and the 'cache misses' is equal to 1.
<img width="1133" alt="Screenshot 2024-10-29 at 16 21 05" src="https://github.com/user-attachments/assets/baef483e-7bc1-47ce-821d-26bd9a56811a">



### Workflow

- [x] Ticket exists in Jira https://uktrade.atlassian.net/browse/GREATUK-1270

### Merging

- [x] This PR can be merged by reviewers.
